### ISSUE TITLE
Use Case: As a researcher I want to be able to define ad hoc properties and use them in a crate without resorting to the verbose PropertyValue thing

### ISSUE THREAD

**Issue by ptsefton** [ResearchObject/ro-crate#71]
I have discussed this in a [blog post](http://ptsefton.com/2020/03/24/semantic_web_hacks.htm).

I think the second solution I provide here may form the basis for a way to satisfy this Use Case.
 
In a nutshell:
-  You define an additional context term (Property) with a URL eg at example.org 
-   Add a definition of the Property to the @graph - and the RO-Crate HTML viewer can then use this to add a definition

**Comment by stain** [comment]
A question is if the property should "match" multiple uses in different documents (e.g. two RO-Crates). 

In the first case it will match as long as not a single part of the definition is changed. Quite immutable!

The second one is overloading example.com so you would be threading in the same space as others talking about sentences, in this sense you are back to the potential conflict as in the `title` example.

How about a local property, like `"@id": "#sentencing"` that then includes the definitions as you have it? Then it is fully defined, even has a identifier. Downside is every mention is then a "new" property, e.g. you would struggle to query over multiple such.

**Comment by ptsefton** [comment]
No - not overloading example.com there's a "namespace" in the URL.

I meant to include an example here - the URL has "criminal characters" in it


```
"@context": [ 
    "https://w3id.org/ro/crate/1.0/context",
    {"sentence": "http://example.com/criminal-characters/sentence"},
  ],
"@graph": [
  {
      "@id": "http://example.com/criminal-characters/sentence",
      "@type": "rdf:Property",
      "rdfs:label": "sentence",
      "rdf:comment": "Penalty imposed by court for criminal conviction. As the data is drawn from prison records, this will usually consist of a specified term of imprisonment and the type of imprisonment conditions, e.g. 6 months hard labour. However, during this historical period it was common for persons convicted of a minor offence to be sentenced to a fine 'with the option' of a prison sentence if they were unable to pay it. After the introduction in Victoria of the Indeterminate Sentences Act in 1907, prisoners who had been declared 'habitual criminals' could also receive an indefinite sentence that meant they were imprisoned until the government authorities determined that they had sufficiently reformed. Some prisoners also faced additional penalties in addition to their prison sentence, such as periods of solitary confinement, in irons or whippings."
      }
    }

```

**Comment by ptsefton** [comment]
Or in this case the researcher has a domain name https://criminalcharacters.com/ we could use https://criminalcharacters.com/vocab/sentence

**Comment by ptsefton** [comment]
I tried using a local name in the context, got this:

**Comment by ptsefton** [comment]
So I think this is the way to add local context. Schema.org has the Property class which is equivalent to the RDF Property Class. There's no `label` or `comment` properties on http://schema.org/Property but it seems obvious to use `name` and `description`.

This is a much more compact and better document option for describing properties used in a crate than the verbose PropertyValue approach.

```
"@context": [ 
    "https://w3id.org/ro/crate/1.0/context",
    {"sentence": "_http://example.com/criminal-characters/sentence"},
  ],
"@graph": [
  {
      "@id": "_http://example.com/criminal-characters/sentence",
      "@type": "Property",
      "name": "sentence",
      "description": "Penalty imposed by court for criminal conviction. As the data is drawn from prison records, this will usually consist of a specified term of imprisonment and the type of imprisonment conditions, e.g. 6 months hard labour. However, during this historical period it was common for persons convicted of a minor offence to be sentenced to a fine 'with the option' of a prison sentence if they were unable to pay it. After the introduction in Victoria of the Indeterminate Sentences Act in 1907, prisoners who had been declared 'habitual criminals' could also receive an indefinite sentence that meant they were imprisoned until the government authorities determined that they had sufficiently reformed. Some prisoners also faced additional penalties in addition to their prison sentence, such as periods of solitary confinement, in irons or whippings."
      }
    }
```

**Comment by ljgarcia** [comment]
I am not sure I understand why you need _ before the sentence link so my comment might be out of tune, please disregard as needed. One thing that I miss on your property definition is the domain and range, unless those are not needed for your use case.
I tried 
```
{
  "@context": [ 
    "https://w3id.org/ro/crate/1.0/context",
    {"sentence": "http://example.com/criminal-characters/sentence"}
  ],
  "@graph": [
    {
      "@id": "http://example.com/criminal-characters/sentence",
      "@type": "Property",
      "name": "sentence",
      "description": "Penalty imposed by court..."
    },
    {
      "@id": "http://my.example.com",
      "@type": "Thing",
      "sentence": "Sentence property range goes here"
    }
  ]
}
```
And it was expanded with JSON playground to 
```
[
  {
    "@id": "http://example.com/criminal-characters/sentence",
    "@type": [
      "http://schema.org/Property"
    ],
    "http://schema.org/description": [
      {
        "@value": "Penalty imposed by court..."
      }
    ],
    "http://schema.org/name": [
      {
        "@value": "sentence"
      }
    ]
  },
  {
    "@id": "http://my.example.com",
    "@type": [
      "http://schema.org/Thing"
    ],
    "http://example.com/criminal-characters/sentence": [
      {
        "@value": "Sentence property range goes here"
      }
    ]
  }
]
```

**Comment by ptsefton** [comment]
@ljgarcia The _ is needed as blank node identifiers are permissable in @context, whereas local URIs and hash-URIs are not. Domain and range might be useful in some applications but I don't thing you have captured that in your example - you have defined an item of @type Thing - would you not use domainIncludes and rangeIncludes properies on the Property item itself?

**Comment by ljgarcia** [comment]
Using {"sentence": "_http://example.com/criminal-characters/sentence"} in the context gave me the same error you reported before
```
jsonld.SyntaxError: Invalid JSON-LD syntax; a @context @id value must be an absolute IRI, a blank node identifier, or a keyword.
```
That is why I thought it was a typo or so. Now that you mention blank node I understand but I still get the error.
I did not tried to include any domain or range as I did not know whether it was relevant for the example provided. The proposed way looks fine as an alternative of PropertyValue but, what would be the disadvantages (if any) of having a local context?

**Comment by ptsefton** [comment]
@ljgarcia  - you're missing the colon.

This works: {"sentence": "_:http://example.com/criminal-characters/sentence"}

But @stain suggested in the may RO-Crate meeting that we use URNS, like:

```
{
  "@context": [ 
    "https://w3id.org/ro/crate/1.0/context",
    {"sentence": "urn:uuid:4dea413c-583d-482f-be5d-1af101600ab7"}
  ],
  "@graph": [
    {
      "@id": "http://example.com/criminal-characters/sentence",
      "@type": "Property",
      "name": "sentence",
      "description": "Penalty imposed by court..."
    }
  ]
}
```

**Comment by stain** [comment]
@ptsefton that's right - however I think it became a lot of UUIDs, you will also have to re-use the same UUID when defining it as prefix-less identifiers in `@id` are not expanded. Since we also want to be able to support ad-hoc enumerations/roles and classes, I think we should always use the prefix. 

I like the idea of using https://schema.org/Class and https://schema.org/Property as they are very lightweight and more flexible than similar rdfs declarations, I'll use these below but I think we should permit a minimal description of just `@type` so that ad-hoc key/values can be represented when the code has no idea what they mean.

## Why not blank nodes?

Here is an example of using `_:blanknodes` which in theory works, but only as long as you stay in JSON-LD or "Generalized RDF:

```json
{
  "@context": [ 
    "http://schema.org/"
  ],
  "@graph": [
    {
      "@id": "_:sentence",
      "@type": "Property",
      "name": "sentence",
      "description": "Penalty imposed by court",
      "domainIncludes": {"@id": "_:Prisoner"}
    },
    {
      "@id": "_:Prisoner",
      "@type": "Class",
      "name": "sentence",
      "description": "Penalty imposed by court"
    },
    {
      "@id":   "#someoneGuilty",
      "@type": "_:Prisoner",
      "_:sentence": "2 nights in jail with bread and water"
    }
  ]
}
```

As pointed out in this [thread](https://lists.w3.org/Archives/Public/uri/2018Nov/0000.html) the problem is that `_:sentence` as a blank node disappears as property from the _N-Quads_, e.g. JSON-LD Playground only outputs:

```turtle
<https://json-ld.org/playground/#someoneGuilty> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> _:b1 .
_:b0 <http://schema.org/description> "Penalty imposed by court" .
_:b0 <http://schema.org/domainIncludes> _:b1 .
_:b0 <http://schema.org/name> "sentence" .
_:b0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Property> .
_:b1 <http://schema.org/description> "Penalty imposed by court" .
_:b1 <http://schema.org/name> "sentence" .
_:b1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Class> .

```

That is we have kept the description of the properties - but sadly they have been renamed to `_:b0` and `_:b1` (as blank nodes permits), as we see in the first line valid as a `@type`, but the triple of the `_:sentence` being bread & water is stripped out as bnodes are not valid RDF properties outside JSON-LD.

## UUIDs to the rescue!

Now here is a brave suggestion - instead of `_` prefix let us make a new `adhoc:` namespace with a fresh UUID per RO-Crate, e.g. `"adhoc": "urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#"`  

If we then use `adhoc:` prefix it makes it clear that the properties are unofficial (they are not imported from anywhere). You avoid double-declaring them both in `@context` and in `@graph` but at the risk of having undeclared properties/typos.

```json
{
  "@context": [ 
    "http://schema.org/",
    { "adhoc": "urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#"
    }
  ],
  "@graph": [
    {
      "@id": "adhoc:sentence",
      "@type": "Property",
      "name": "sentence",
      "description": "Penalty imposed by court",
      "domainIncludes": {"@id": "adhoc:Prisoner"}
    },
    {
      "@id": "adhoc:Prisoner",
      "@type": "Class",
      "name": "sentence",
      "description": "Penalty imposed by court"
    },
    {
      "@id":   "#someoneGuilty",
      "@type": "adhoc:Prisoner",
      "adhoc:sentence": "2 nights in jail with bread and water"
    }
  ]
}
```

Now both the names after `:` is preserved, along with the UUID for both properties and classes:

```turtle
<https://json-ld.org/playground/#someoneGuilty> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#Prisoner> .
<https://json-ld.org/playground/#someoneGuilty> <urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#sentence> "2 nights in jail with bread and water" .

<urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#Prisoner> <http://schema.org/description> "Penalty imposed by court" .
<urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#Prisoner> <http://schema.org/name> "sentence" .
<urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#Prisoner> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Class> .

<urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#sentence> <http://schema.org/description> "Penalty imposed by court" .
<urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#sentence> <http://schema.org/domainIncludes> <urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#Prisoner> .
<urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#sentence> <http://schema.org/name> "sentence" .
<urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#sentence> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Property> .
```

Now the only slightly edgy thing about this is using `#` on the `urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389` - fragments on URNs are undefined as they are not (generally) resolvable as a file -  but as our UUID is meant to represent our **adhoc ontology** - and as the UUID owner could decide that its fragment work as above. See this [thread](https://lists.w3.org/Archives/Public/uri/2011Mar/0009.html))

### Reuse of ontology

It could be someone finds a nice ad-hoc terms in an RO-Crate and want to use it again, perhaps along new ad-hoc terms under a different UUID. In this case the old term not strictly ad-hoc and so should be declared in the `@context` as some kind of `prison` ontology:

```json
{
  "@context": [ 
    "http://schema.org/",
    { "adhoc": "urn:uuid:84b6db45-3ea9-40e3-97ab-d2fb8306486a#",
      "prison": "urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#"
    }
  ],
  "@graph": [
     {
      "@id":   "urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#sentence",
      "name": "sentence",
      "url": {"@id": "http://example.com/the-ro-that-made-771176f2"}
    },
    {
      "@id":   "urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#Prisoner",
      "url": {"@id": "http://example.com/the-ro-that-made-771176f2"}
    },
    {
      "@id":   "#appleThief",
      "@type": "prison:Prisoner",
      "prison:sentence": "3 nights in jail with bread and water",
      "adhoc:because": "Stole lots of apples"
    }
  ]
}
```

Now `prison:sentence` is similar to the Bibo example in [Extending RO-Crate](https://researchobject.github.io/ro-crate/1.0/#extending-ro-crate) and so we provide only a `url` to the RO-Crate that introduced it, perhaps a `name` to help users.

Triples would then have two ad-hoc namespaces:

```
<https://json-ld.org/playground/#appleThief> <urn:uuid:771176f2-e081-4fb7-8d0f-e36b1617c389#sentence> "3 nights in jail with bread and water" .

<https://json-ld.org/playground/#appleThief> <urn:uuid:84b6db45-3ea9-40e3-97ab-d2fb8306486a#because> "Stole lots of apples" .
```



### Middle-ground: UUIDs in http

A "kinder" workaround would be to make a http URI as a permalink that has not yet been registered: ` "adhoc": "https://w3id.org/uuid/771176f2-e081-4fb7-8d0f-e36b1617c389/"`  so then properties would become https://w3id.org/uuid/771176f2-e081-4fb7-8d0f-e36b1617c389/Prisoner etc.

We could add an empty `uuid/` folder under https://github.com/perma-id/w3id.org - in theory anyone could then later register the UUID there to "claim" an ontology or even just a term, and redirect to an ontology page. 

However I am not sure if that good or bad as there would be no way of knowing who minted the UUID! (bring in your network card to prove its MAC address!)

### Lots of UUIDs

In this approach, as suggested by @ptsefton, every ad-hoc term has a fresh UUID. There is thus no implication that these terms are grouped in any way in a particular ontology, but this could make it easier to re-use a term across RO-Crate.

**Comment by stain** [comment]
If we agree on what approach to go for with the ad-goc properties I can make the changes in a pull request for RO-Crate 1.1.
 
Views from @ptsefton @ljgarcia @alaninmcr @AlasdairGray @scapella @markwilkinson @dgarijo et al?

**Comment by marcolarosa** [comment]
This has been implemented in Describo via the profiles. That is, a user can't just add any property via the GUI but the profile author can add their own properties, with definitions, into the profile and they will be added in to the crate.

The profile definition for a custom property is described @ https://github.com/UTS-eResearch/describo/wiki/dsp-input-definition#defining-your-own-properties.

This results in the following in the graph:
```
{
  "@context": [
    "https://w3id.org/ro/crate/1.0/context",
    {
      "@vocab": "https://schema.org/",
      "orthographicNotes": "_:orthographicNotes"
    }
  ],
  "@graph": [
    {
      "@id": "/ro-crate-metadata.json",
    ....
    },
    {
      "@type": "Dataset",
      "name": "my crate",
      "description": "something or other.w",
      "orthographicNotes": "Some content",
      "@id": "./"
    },
    {
      "@id": "_:orthographicNotes",
      "@type": "Property",
      "name": "orthographicNotes",
      "description": ""
    }
  ]
}
```

Notice that the property `orthographicNotes` in the root dataset has a definition in the context that points to the entry in the graph with the definition.

**Comment by marcolarosa** [comment]
If you want to see this for yourself load in the paradisec profile which has a number of custom properties. Use https://raw.githubusercontent.com/UTS-eResearch/describo/master/src/components/profiles/paradisec.describo.profile.json on the profile selection page in the box "Load from a URL"

**Comment by dgarijo** [comment]
@stain I am very sorry, for some reason I had not seen the previous proposals with urns.

In short:
- I would **not** recommend adopting blank nodes. If the scope is for developers to adopt our approach, I am not sure blank nodes would help our cause. I haven't met a developer who was happy with blank nodes, and there is even a debate on the semantic web community on whether bnodes should be used or not.
- Some of the proposals seem a little complicated to me. For example, if developers need to reuse properties defined in another rocrate, why not create a common vocabulary and make it available? Are we assuming they wouldn't know how to do that but they would be comfortable trying to disambiguate terms of different rocrates? 
- I am not against using urns for new properties. But if a property exists within the scope of a ro crate only, does it really need to be resolved? Why use a urn at all? For example, what if we enable 2 w3ids: one for common properties (https://w3id.org/roCrate/extension/ (or adhoc, as you propose)); and one for properties that i am using in my rocrate but I don't really care where they are resolved to (https://w3id.org/roCrate/example/). The condition would be that if you have properties in common for ro crates, please consider defining them in /extension. Otherwise just do httsp://w3id.org/roCrate/example/uuid/771176f2-e081-4fb7-8d0f-e36b1617c389/. This would mean that, as a user, that property is defined within the rocrate itself.

**Comment by dgarijo** [comment]
@stain @ptsefton I have created https://github.com/ResearchObject/ro-terms
Please let me know what you think about my proposal, which is collect terms with a CSV. I added 2 terms.
If you like it, we can do a small script to trnasform the CSV to JSON-LD, turtle or other serialization.
But this is a simple way of seeing what it's there without requiring to edit JSON or OWL.

**Comment by dgarijo** [comment]
I made the https://w3id.org/ro/terms point to the CSV. However, when we agree we could enable content negotiation and point there just for the text representation

**Comment by ptsefton** [comment]
Sorry I didn't comment before! 

I was thinking about letting people add HTML or markdown pages as definitions but it seems like # urls don't actually work in github - so your CSV idea looks like a good solution as long as we generate HTML from it so URLs resolve. I would suggest that we might want to add a column for Class / Property rather than having to guess that from the first letter. 

Also I think we do need to have a namespace so people can claim their own "area" and we don't get overlap

**Comment by ptsefton** [comment]
One question about the CSV - what's the difference between term and label? I suggest we stick to something that maps to what they do at Schema.org:

https://schema.org/version/latest/schemaorg-current-http.jsonld

**Comment by ptsefton** [comment]
@dgarijo Maybe use this form for the CSV - not sure if this is documented anywhere.

https://schema.org/version/latest/schemaorg-current-http-types.csv

**Comment by ptsefton** [comment]
@dgarijo I have made a branch with some draft text for your consideration.

https://github.com/ResearchObject/ro-crate/blob/issue/71/docs/1.1-DRAFT/index.md#adding-new-or-ad-hoc-vocabulary-terms

**Comment by dgarijo** [comment]
@ptsefton,
Some answers:
-  The difference between term and label: The term is the term id after your vocabulary namespace, while the label is a human-readable label of the term. For example the term "presidentOf" could have the label "is president of". Or the term "P123" could have the label "publisher" (as in wikidata). If you think that adopting the schema.org types and properties is easier, we can do it. Having 2 separate documents can be problematic: What would happen when I resolve the namespace? A solution for this would be to have 2 separate csvs, but point to a json-ld file which is created automatically (I like this approach)

- Right now we only have https://w3id.org/ro/terms, but we can have https://w3id.org/ro/terms/ptsefton redirect to your folder in the vocabulary repository, as long as we name the csv/json in a similar structure. We should have an example or two about this.

**Comment by stain** [comment]
I agree with using folders for each namespace - the whole point of this was to gather potentially clashing terms that may also come from some third-party system. If it's one big CSV file then updates will be tricky, and one change could break the whole list.

But I think we should encourage the "namespace" to be very close to the thing it is describing as in Peter's `criminalcharacters` example - not by person or organization (in which case they should try to use their own domain names).

In https://bioschemas.org/ we did something similar to this - starting with Google Spreadsheets. 

I think generating JSON-LD would be nice (perhaps with a `@context` that can be imported as `https://w3id.org/ro/namespace`), but HTML-rendering for humans even better. This is after all going to be ad-hoc and not have a particularly semantic structure.

**Comment by stain** [comment]
OK, I merged @ptsefton's suggestion in #90 (agreed in joint call). Now we just need to reflect the namespace folder thing on https://github.com/ResearchObject/ro-terms - we don't need to wait for CSV-processing-magic-script before we push RO-Crate 1.1 spec.

Current draft: https://www.researchobject.org/ro-crate/1.1-DRAFT/

So then I will also go ahead and prepare a release candidate of 1.1 that we can vote on for the RO-Crate call next week Thursday.

**Comment by dgarijo** [comment]
Thanks, @stain 
Before the call tomorrow I will try to improve a little on the redirection, reserving namespaces, etc. Maybe we can discuss briefly the draft in the call too.
